lib/ProductOpener/APIProductServices.pm

```diff
@@ -0,0 +1,180 @@
+# This file is part of Product Opener.
+#
+# Product Opener
+# Copyright (C) 2011-2023 Association Open Food Facts
+# Contact: contact@openfoodfacts.org
+# Address: 21 rue des Iles, 94100 Saint-Maur des Fossés, France
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+=head1 NAME
+
+ProductOpener::APIProductServices - Microservices to enrich a product object
+
+=head1 DESCRIPTION
+
+This module implements a microservice API for operations done on an product object.
+
+Applications can send product data (for instance a nested list of ingredients),
+ask for one or more services to be executed on the input product data
+(for instance computing the min, max and estimated percentages of each ingredient),
+and get back resulting product data (possibly filtered to get only specific fields back).
+
+=head2 INTERFACE
+
+=head3 Request
+
+The Routing.pm and API.pm module offer an HTTP interface of this form:
+POST /api/v3/product_services/[comma separated list of services]
+
+The POST body is a JSON hash with those fields:
+
+=head4 product
+
+A product object
+
+=head4 field
+
+A comma separated list of fields to return. If empty, all available fields
```

Comment:
Could this also be an array instead of a comma separated list?

```json
{
    "fields": ["code", "categories"]
}
```

will probably feel more natural to people working with JSON than

```json
{
    "fields": "code,categories"
}
```

Reply:
Good idea.

Reply:
I'd prefer that each API return just the fields that it modifies / generates

Reply:
I agree with john, returning by default only the modified/generated fields with an option to return all fields would be better imo

Reply:
In API v3 product WRITE, we already have a special "updated" value for fields: only updated fields are returned. We can do the same here, and make "updated" the default instead of "all".

https://github.com/openfoodfacts/openfoodfacts-server/blob/main/docs/api/ref/requestBodies/fields_tags_lc.yaml

Reply:
By default the API now only returns the created or updated fields. For structures that are updated (like when we add percent fields to the ingredients structure), I return the whole structure.